src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableExtensions.cs

```diff
@@ -103,15 +155,25 @@ internal static T[] ToArray<T>(this IEnumerable<T> sequence, int count)
             Requires.NotNull(sequence, nameof(sequence));
             Requires.Range(count >= 0, nameof(count));
 
+            if (count == 0)
+            {
+                return ImmutableArray<T>.Empty.array;
+            }
+
             T[] array = new T[count];
-            int i = 0;
-            foreach (var item in sequence)
+
+            if (!sequence.TryCopyTo(array, 0))
             {
-                Requires.Argument(i < count);
-                array[i++] = item;
+                int i = 0;
+                foreach (var item in sequence)
+                {
+                    Requires.Argument(i < count);
+                    array[i++] = item;
+                }
+
+                Requires.Argument(i == count);
             }
 
-            Requires.Argument(i == count);
```

Comment:
With your change, we've lost the validation that the source enumerable's length matched the length of the allocated array. So now we just have to fully trust the `count` parameter. Yet you've left the validation of it inside the `if` block, so it seems we both implicitly trust it, and we don't. It seems inconsistent.